src/main/java/com/fasterxml/jackson/databind/deser/BeanDeserializerBase.java

```diff
@@ -177,8 +179,14 @@ public abstract class BeanDeserializerBase
      * Note that this is <b>only needed</b> for polymorphic types,
      * that is, when the actual type is not statically known.
      * For other types this remains null.
+     * The map type changed in 2.18 (from HashMap to ConcurrentHashMap)
      */
-    protected transient HashMap<ClassKey, JsonDeserializer<Object>> _subDeserializers;
+    protected transient ConcurrentHashMap<ClassKey, JsonDeserializer<Object>> _subDeserializers;
+
+    /**
+     * Lock used to prevent multiple creation of _subDeserializers map
+     */
+    private final ReentrantLock _subDeserializersLock = new ReentrantLock();
```

Comment:
This now adds ReentrantLock for every `BeanDeserializer` instance, regardless of whether polymorphic handling is enabled. PM handling likely being a small minority of all usage.